### ISSUE TITLE
Google Container Registry

### ISSUE THREAD

**Issue by DazWilkin** [engineerd/wasm-to-oci#9]
I found this solution via DeisLabs' [Krustlet](https://github.com/deislabs/krustlet)

I'd like to use [Google Container Registry (GCR)](https://cloud.google.com/container-registry) with `wasm-to-oci` but I'm struggling with the auth (pass-through to [ORAS](https://github.com/deislabs/oras)).

[GCR supports OCI](https://cloud.google.com/container-registry/docs/image-formats#oci_format) and, using ORAS, I'm able:

+ [authenticate](https://github.com/deislabs/oras/issues/157#issuecomment-623594613) using a (Google) access token (and demonstrate that GCR support OCI containers).
+ push|pull images to a GCR repo.

```bash
REGISTRY="gcr.io"

gcloud auth print-access-token \
| oras login -u oauth2accesstoken --password-stdin https://${REGISTRY}

more ./artifact.txt 
Hello Freddie!

oras push gcr.io/${PROJECT}/hello-artifact:v1 ./artifact.txt
Uploading 0060e7e892de artifact.txt
Pushed gcr.io/${PROJECT}/hello-artifact:v1
Digest: sha256:c330be5cf8bbc34f87db380aaba72e774d7ee1064085b3e2ec2b071be292e2a2

rm artifact.txt 

oras pull gcr.io/${PROJECT}/hello-artifact:v1
Downloaded 0060e7e892de artifact.txt
Pulled gcr.io/${PROJECT}/hello-artifact:v1
Digest: sha256:c330be5cf8bbc34f87db380aaba72e774d7ee1064085b3e2ec2b071be292e2a2

cat ./artifact.txt 
Hello Freddie!
```

If I could understand how to reflect this authentication using `wasm-to-oci`, I think GCR could be added to your list of supported registries.

**Comment by bacongobbler** [comment]
Have you tried the same steps, only replacing `oras pull` and `oras push` with `wasm-to-oci pull|push`? Under the covers, wasm-to-oci [uses oras](https://github.com/engineerd/wasm-to-oci/blob/7c307e8e8c52a9eb2e17038424d20af3e46812f7/pkg/oci/pull.go#L14-L22) for its pull and push functionality, so it''s entirely possible that an `oras login` followed by a `wasm-to-oci` push could work.

Let us know how your research goes.

**Comment by DazWilkin** [comment]
Wow! That was a quick reply! ;-) Thank you!

User error, apologies :-(

I'd mistyped the value of `${PROJECT}` with `wasm-to-oci`

Success!

```bash
gcloud auth print-access-token \
| go run ./cmd/oras login -u oauth2accesstoken --password-stdin https://gcr.io
Login Succeeded

go run ./cmd/oras pull gcr.io/${PROJECT}/hello-artifact:v1
Downloaded 0060e7e892de artifact.txt
Pulled gcr.io/${PROJECT}/hello-artifact:v1
Digest: sha256:c330be5cf8bbc34f87db380aaba72e774d7ee1064085b3e2ec2b071be292e2a2

go run ./cmd push testdata/hello.wasm gcr.io/${PROJECT}/hello-wasm:v1
INFO[0003] Pushed: gcr.io/dazwilkin-200504-krustlet/hello-wasm:v1 
INFO[0003] Size: 1624962                                
INFO[0003] Digest: sha256:9c82cbe576ee947c00435ac8053a800a1969f4757ae4a81f870f714674afc91a 
```
And:
![image](https://user-images.githubusercontent.com/1022400/81001272-55dc5700-8dfc-11ea-9f12-1e2f699dac29.png)

How may I auth directly using `wasm-to-oci` without having to `oras login`?

**Comment by radu-matei** [comment]
@DazWilkin `wasm-to-oci` reuses the credentials from `~/.docker/config.json`, so if you are logged in to a registry using any CLI tool (`docker`, `oras`, or `gcloud`), it should work.

Do you think adding an `wasm-to-oci login` command would be useful?
I purposely tried to avoid that since it would just reimplement what those tools are doing, but I am open to suggestions.

**Comment by DazWilkin** [comment]
@radu-matei thanks for asking.

I think it's likely not necessary but it *may* be helpful to more clearly document the auth mechanism.

I began at [`Krustlet`](https://github.com/deislabs/krustlet) which references `wasm-to-oci` which references [`ORAS`](https://github.com/deislabls/oras) so there three distinct novelties for me and I was unfamiliar with the concept of OCI registries.

I'm also running docker as a Snap and this complicates auth (because it doesn't default to `{$HOME}/.docker/config.json`).

Once I realized, that I just needed to just `docker login` to GCR, I was in the clear.

It would be good to see GCR referenced by both tools as a registry that supports OCI ;-)
